bot/tests/processors/test_pennychat.py

```diff
@@ -150,3 +165,177 @@ def id_mock(user_id, slack_client):
+        starting_role=Participant.ORGANIZER,
+        can_attend=False,
+        expected_final_role=Participant.ORGANIZER,
+        expected_organizer_notified=False,
+    ),
+    dict(
+        msg="if invitee can attend then her role becomes ATTENDEE and the organizer is notified",
+        starting_role=Participant.INVITEE,
+        can_attend=True,
+        expected_final_role=Participant.ATTENDEE,
+        expected_organizer_notified=True,
+    ),
+    dict(
+        msg="if invitee not can attend then her role becomes INVITED_NONATTENDEE and the organizer is notified",
+        starting_role=Participant.INVITEE,
+        can_attend=False,
+        expected_final_role=Participant.INVITED_NONATTENDEE,
+        expected_organizer_notified=True,
+    ),
+    dict(
+        msg=(
+            "if an attendee decides they can't attend, then their role is changed to "
+            "INVITED_NONATTENDEE and the organizer is notified"
+        ),
+        starting_role=Participant.ATTENDEE,
+        can_attend=False,
+        expected_final_role=Participant.INVITED_NONATTENDEE,
+        expected_organizer_notified=True,
+    ),
+    dict(
+        msg=(
+            "if an invited_nonattendee desides to attend, then their role is changed to "
+            "ATTENDEE and the organizer is notified"
+        ),
+        starting_role=Participant.INVITED_NONATTENDEE,
+        can_attend=True,
+        expected_final_role=Participant.ATTENDEE,
+        expected_organizer_notified=True,
+    ),
+    dict(
+        msg=(
+            "if an attendee choose can_attend again, then the participation is "
+            "not changed and the organizer is not notified"
+        ),
+        starting_role=Participant.ATTENDEE,
+        can_attend=True,  # they've already said they'll be there and they say it again
+        expected_final_role=Participant.ATTENDEE,
+        expected_organizer_notified=False,
+    ),
+    dict(
+        msg=(
+            "if an nonattendee chooses can_not_attend again, "
+            "then the participation is not changed and the organizer is not notified"
+        ),
+        starting_role=Participant.INVITED_NONATTENDEE,
+        can_attend=False,
+        expected_final_role=Participant.INVITED_NONATTENDEE,
+        expected_organizer_notified=False,
+    ),
+    dict(
+        msg="if non-participant can attend then attendance participation is created and notification is delivered",
+        starting_role=None,
+        can_attend=True,
+        expected_final_role=Participant.ATTENDEE,
+        expected_organizer_notified=True,
+    ),
+    dict(
+        msg="if non-participant can not attend then no participation is created and no notification is delivered",
+        starting_role=None,
+        can_attend=False,
+        expected_final_role=None,
+        expected_organizer_notified=False,
+    ),
+]
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize(",".join(testdata[0].keys()), [td.values() for td in testdata])
```

Comment:
Ok, this is awesome. Wasn't aware of this before!

Reply:
yeah! me neither - super useful